File: src/main-dokka/src/wemi/documentation/impl/DokkaInterfaceImpl.java
package wemi.documentation.impl;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.dokka.DokkaConfiguration;
import org.jetbrains.dokka.DokkaConfigurationImpl;
import org.jetbrains.dokka.DokkaGenerator;
import org.jetbrains.dokka.DokkaModuleDescriptionImpl;
import org.jetbrains.dokka.DokkaSourceSetID;
import org.jetbrains.dokka.DokkaSourceSetImpl;
import org.jetbrains.dokka.ExternalDocumentationLinkImpl;
import org.jetbrains.dokka.PackageOptionsImpl;
import org.jetbrains.dokka.Platform;
import org.jetbrains.dokka.SourceLinkDefinitionImpl;
import org.jetbrains.dokka.utilities.DokkaLogger;
import org.slf4j.Logger;
import org.slf4j.Marker;
import wemi.documentation.DokkaInterface;
import wemi.documentation.DokkaOptions;

import java.net.MalformedURLException;
import java.net.URL;
import java.nio.file.Path;
import java.util.Collection;
import java.util.Collections;
import java.util.stream.Collectors;

/**
 * {@link DokkaInterface} implementation, DO NOT TOUCH FROM ELSEWHERE THAN {@link wemi.KeyDefaults.ArchiveDokkaInterface}!!!
 *
 * In Java, because of https://youtrack.jetbrains.com/issue/KTIJ-530
 */
public class DokkaInterfaceImpl implements DokkaInterface {

	private static URL toUrl(String url) {
		try {
			return new URL(url);
		} catch (MalformedURLException e) {
			throw new RuntimeException("Failed to create URL from '"+url+"'", e);
		}
	}

	@Override
	public void execute(
			@NotNull Collection<? extends Path> classpath, @NotNull Path outputDirectory,
			@Nullable Path packageListCache, @NotNull DokkaOptions options,
			@NotNull Logger logger, @Nullable Marker loggerMarker) {

		DokkaConfiguration config = new DokkaConfigurationImpl(
				options.getModuleName(),
				options.getModuleVersion(),
				outputDirectory.toFile(),
				packageListCache != null ? packageListCache.toFile() : null,
				options.getOffline(),
				Collections.singletonList(new DokkaSourceSetImpl(
						options.getModuleName(),
						new DokkaSourceSetID(options.getModuleName(), options.getModuleName()), // ?
						classpath.stream().map(Path::toFile).collect(Collectors.toList()),
						options.getSourceRoots().stream().map(root -> root.getDir().toFile()).collect(Collectors.toSet()),
						Collections.emptySet(),
						options.getSampleRoots().stream().map(Path::toFile).collect(Collectors.toSet()),
						options.getIncludes().stream().map(Path::toFile).collect(Collectors.toSet()),
						options.getIncludeNonPublic(),
						options.getReportNotDocumented(),
						options.getSkipEmptyPackages(),
						options.getSkipDeprecated(),
						options.getJdkVersion(),
						options.getSourceLinks().stream().map(s -> new SourceLinkDefinitionImpl(s.getDir().toAbsolutePath().toString(), toUrl(s.getUrl()), s.getUrlSuffix())).collect(Collectors.toSet()),
						options.getPerPackageOptions().stream().map(p -> new PackageOptionsImpl(p.getPrefix(), p.getIncludeNonPublic(), p.getReportNotDocumented(), p.getSkipDeprecated(), false)).collect(Collectors.toList()),
						options.getExternalDocumentationLinks().stream().map(e -> new ExternalDocumentationLinkImpl(toUrl(e.getUrl()), toUrl(e.getPackageListUrl() != null ? e.getPackageListUrl() : e.getUrl()+"/package-list"))).collect(Collectors.toSet()),
						"",
						"",
						options.getNoStdlibLink(),
						options.getNoJdkLink(),
						Collections.emptySet(),
						Platform.jvm // For now
				)),
				Collections.emptyList(),// TODO(jp): The plugins are actually vital
				Collections.emptyList(),
				Collections.singletonList(new DokkaModuleDescriptionImpl(options.getModuleName(), outputDirectory.toFile(), options.getSourceRoots().stream().map(root -> root.getDir().toFile()).collect(Collectors.toSet()))),
				false
		);

		final DokkaGenerator gen = new DokkaGenerator(config, new DokkaLogger() {

			private int warningsCount = 0;
			private int errorsCount = 0;

			@Override
			public int getWarningsCount() {
				return warningsCount;
			}

			@Override
			public void setWarningsCount(int i) {
				warningsCount = i;
			}

			@Override
			public int getErrorsCount() {
				return errorsCount;
			}

			@Override
			public void setErrorsCount(int i) {
				errorsCount = i;
			}

			@Override
			public void debug(@NotNull String s) {
				logger.debug(loggerMarker, "{}", s);
			}

			@Override
			public void info(@NotNull String s) {
				logger.info(loggerMarker, "{}", s);
			}

			@Override
			public void progress(@NotNull String s) {
				logger.debug(loggerMarker, "Progress: {}", s);
			}

			@Override
			public void warn(@NotNull String s) {
				warningsCount++;
				logger.warn(loggerMarker, "{}", s);
			}

			@Override
			public void error(@NotNull String s) {
				errorsCount++;
				logger.error(loggerMarker, "{}", s);
			}
		});

		gen.generate();
	}
}

File: src/main/kotlin/wemi/documentation/impl/DokkaInterfaceImpl.kt
package wemi.documentation.impl

import org.jetbrains.dokka.DocumentationOptions
import org.jetbrains.dokka.DokkaConfiguration
import org.jetbrains.dokka.DokkaGenerator
import org.jetbrains.dokka.DokkaLogger
import org.slf4j.Logger
import org.slf4j.Marker
import wemi.documentation.DokkaInterface
import wemi.documentation.DokkaOptions
import wemi.util.LocatedFile
import wemi.util.absolutePath
import java.net.URL
import java.nio.file.Path

/**
 * [DokkaInterface] implementation, DO NOT TOUCH FROM ELSEWHERE THAN [wemi.KeyDefaults.Dokka]!!!
 */
@Suppress("unused")
internal class DokkaInterfaceImpl : DokkaInterface {

    override fun execute(sources: Collection<LocatedFile>,
                         classpath: Collection<Path>,
                         outputDirectory:Path,
                         packageListCache:Path?,
                         options:DokkaOptions,
                         logger: Logger,
                         loggerMarker: Marker?) {

        val sourceRoots = HashSet<Path>()
        for (source in sources) {
            sourceRoots.add(source.root)
        }

        val gen = DokkaGenerator(object : DokkaLogger {

            override fun info(message: String) {
                logger.info(loggerMarker, "{}", message)
            }

            override fun warn(message: String) {
                logger.warn(loggerMarker, "{}", message)
            }

            override fun error(message: String) {
                logger.error(loggerMarker, "{}", message)
            }
        },
                classpath.map { it.absolutePath },
                sourceRoots.map { root ->
                    val rootPath = root.absolutePath
                    object : DokkaConfiguration.SourceRoot {
                        override val path: String
                            get() = rootPath
                        override val platforms: List<String>
                            get() = emptyList()

                    }
                },
                options.sampleRoots.map { it.absolutePath },
                options.includes.map { it.absolutePath },
                options.moduleName,
                DocumentationOptions(outputDirectory.absolutePath, options.outputFormat,
                        sourceLinks = options.sourceLinks.map { s ->
                            object : DokkaConfiguration.SourceLinkDefinition {
                                override val lineSuffix: String?
                                    get() = s.urlSuffix
                                override val path: String
                                    get() = s.dir.absolutePath
                                override val url: String
                                    get() = s.url
                            }
                        },
                        jdkVersion = options.jdkVersion,
                        skipDeprecated = options.skipDeprecated,
                        skipEmptyPackages = options.skipEmptyPackages,
                        reportUndocumented = options.reportNotDocumented,
                        impliedPlatforms = options.impliedPlatforms,
                        perPackageOptions = options.perPackageOptions.map { o ->
                            object : DokkaConfiguration.PackageOptions {
                                override val prefix: String
                                    get() = o.prefix
                                override val includeNonPublic: Boolean
                                    get() = o.includeNonPublic
                                override val reportUndocumented: Boolean
                                    get() = o.reportNotDocumented
                                override val skipDeprecated: Boolean
                                    get() = o.skipDeprecated
                            }
                        },
                        externalDocumentationLinks = options.externalDocumentationLinks.map { e ->
                            val docUrl = URL(e.url)
                            val docPackageListUrl:URL
                            if (e.packageListUrl == null) {
                                docPackageListUrl = URL(docUrl, "package-list")
                            } else {
                                docPackageListUrl = URL(e.packageListUrl)
                            }

                            object : DokkaConfiguration.ExternalDocumentationLink {
                                override val packageListUrl: URL
                                    get() = docPackageListUrl
                                override val url: URL
                                    get() = docUrl
                            }
                        },
                        noStdlibLink = options.noStdlibLink,
                        cacheRoot = packageListCache?.absolutePath
                )
        )

        gen.generate()
    }

}